trieste/space.py

```diff
@@ -32,6 +34,180 @@
+    """
+    A wrapper class for nonlinear constraints on variables. The constraints expression is of the
+    form::
+
+        lb <= fun(x) <= ub
+
+    :param fun: The function defining the nonlinear constraints; with input shape [..., D] and
+        output shape [..., M], where M is the number of constraints.
+    :param lb: The lower bound of the constraint. Should be a scalar or of shape [M].
+    :param ub: The upper bound of the constraint. Should be a scalar or of shape [M].
+    :param keep_feasible: Keep the constraints feasible throughout optimization iterations if this
+        is `True`.
+    """
+
+    def __init__(
+        self,
+        fun: Callable[[TensorType], TensorType],
+        lb: Sequence[float] | TensorType,
+        ub: Sequence[float] | TensorType,
+        keep_feasible: bool = False,
+    ):
+
+        # Implement caching to avoid calling the constraint function multiple times to get value
+        # and gradient.
+        def _constraint_value_and_gradient(x: TensorType) -> Tuple[TensorType, TensorType]:
+            val, grad = tfp.math.value_and_gradient(fun, x)
+            return tf.cast(val, dtype=x.dtype), tf.cast(grad, dtype=x.dtype)
+
+        cache_x: TensorType = tf.constant([])
+        cache_f: TensorType = tf.constant([])
+        cache_df_dx: TensorType = tf.constant([])
+
+        def val_fun(x: TensorType) -> TensorType:
+            nonlocal cache_x, cache_f, cache_df_dx
+            if not np.array_equal(x, cache_x):
+                cache_f, cache_df_dx = _constraint_value_and_gradient(x)
+                cache_x = x
+            return cache_f
+
+        def jac_fun(x: TensorType) -> TensorType:
+            nonlocal cache_x, cache_f, cache_df_dx
+            if not np.array_equal(x, cache_x):
+                cache_f, cache_df_dx = _constraint_value_and_gradient(x)
+                cache_x = x
+            return cache_df_dx
+
+        self._orig_fun = fun  # Used for constraints equality check.
+        super().__init__(val_fun, lb, ub, jac=jac_fun, keep_feasible=keep_feasible)
+
+    def residual(self, points: TensorType) -> TensorType:
+        """
+        Calculate the residuals between the constraint function and its lower/upper limits.
+
+        For example:
+
+        >>> points = tf.constant([[-1.0, 0.4], [-1.0, 0.6], [0.0, 0.4]], dtype=DEFAULT_DTYPE)
+        >>> nlc = NonlinearConstraint(lambda x: x[..., 0] - tf.math.sin(x[..., 1]), -1.4, 1.9)
+        >>> nlc.residual(points)
+        <tf.Tensor: shape=(3, 2), dtype=float64, numpy=
+        array([[ 0.01058163,  3.28941832],
+               [-0.1646425 ,  3.46464245],
+               [ 1.01058163,  2.28941832]])>
+
+
+        :param points: The points to calculate the residuals for, with shape [..., D].
+        :return: A tensor containing the lower and upper residual values with shape [..., M*2].
+        """
+        non_d_axes = tf.ones_like(tf.shape(points))[:-1]
```

Comment:
Maybe add some shape/rank check on points?

Reply:
Added in 30a17e7